web/src/postprocess_content.ts

```diff
@@ -66,6 +66,38 @@ export function postprocess_content(html: string): string {
             elt.removeAttribute("target");
         }
 
+        if (!elt.parentElement?.classList.contains("message_inline_image")) {
+            // For non-media (images, video) user uploads, the following block
+            // ensures that the title attribute always displays the filename,
+            // as a security measure.
+            let title: string;
+            let legacy_title: string;
+            if (
+                url.origin === window.location.origin &&
+                url.pathname.startsWith("/user_uploads/")
+            ) {
+                // We add the word "download" to make clear what will
+                // happen when clicking the file.  This is particularly
+                // important in the desktop app, where hovering a URL does
+                // not display the URL like it does in the web app.
+                title = legacy_title = $t(
+                    {defaultMessage: "Download {filename}"},
+                    {
+                        filename: decodeURIComponent(
+                            url.pathname.slice(url.pathname.lastIndexOf("/") + 1),
+                        ),
+                    },
+                );
+            } else {
+                title = url.toString();
+                legacy_title = href;
+            }
+            elt.setAttribute(
+                "title",
+                ["", legacy_title].includes(elt.title) ? title : `${title}\n${elt.title}`,
+            );
+        }
```

Comment:
Moving this out of the `else` clause makes the security assertion that we never trust `title` weaker. I think an effective way to handle this detail would be to move `const title = elt.getAttribute("title")` above this block, and pop it off it present unconditionally.